packages/std/src/addresses.rs

```diff
@@ -42,49 +53,111 @@ impl Addr {
     /// let address = Addr::unchecked("foobar");
     /// assert_eq!(address, "foobar");
     /// ```
-    pub fn unchecked<T: Into<String>>(input: T) -> Addr {
-        Addr(input.into())
+    pub const fn unchecked(input: &str) -> Addr {
+        let input_bytes = input.as_bytes();
+        let mut data = [0u8; ADDR_MAX_LENGTH];
+        let mut i = 0;
+        while i < input_bytes.len() {
+            data[i] = input_bytes[i];
+            i += 1;
+        }
+
+        Addr(AddrInner {
+            data,
+            length: input_bytes.len(),
+        })
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0.data[..self.0.length]
+    }
+
+    pub fn as_str(&self) -> &str {
+        unsafe { str::from_utf8_unchecked(self.as_bytes()) }
     }
 }
 
 impl fmt::Display for Addr {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", &self.0)
+        write!(f, "{}", self.as_str())
+    }
+}
+
+impl Serialize for Addr {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        serializer.serialize_str(&self.as_str())
+    }
+}
+
+impl<'de> Deserialize<'de> for Addr {
+    fn deserialize<D>(deserializer: D) -> Result<Addr, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_str(AddrVisitor)
+    }
+}
+
+struct AddrVisitor;
+
+impl<'de> de::Visitor<'de> for AddrVisitor {
+    type Value = Addr;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("Human readable address string")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        if v.len() > ADDR_MAX_LENGTH {
```

Comment:
Nice to assert this here.
Should we assert this in `unchecked` as well? Or would that violate the `const fn` (you could just panic there, but it would ensure nothing gets chopped off by accident)